text/0328-polyglot-assert.md

```diff
@@ -0,0 +1,336 @@
+will be marked as deprecated.
+
+When the module is ready for prime time, we will publish the APIs without the 'Beta'
+suffix and mark all 'Beta' APIs as deprecated.
+
+## Internal FAQ
+
+### Why are we doing this?
+
+We have received a lot of requests and feedback from users that they would like to use
+our 'assert' library in the languages of their choice, primarily Python users.
+
+### Why should we _not_ do this?
+
+None.
+
+### What changes are required to enable this change?
+
+The high level design is to create a new jsii module in the AWS CDK monorepo -
+`@aws-cdk/assertions` - that exposes the APIs proposed above.
+
+The implementation of this module would simply be a scaffold around the existing
+`assert` library.
+
+To build this scaffold, we will 'vendor in' the libraries `assert-internal`,
+`cloudformation-diff` and `cfnspec`.
+
+See [Appendix A](#appendix-a---detailed-design) for the design and rationale.
+
+Prototype / Implementation: <https://github.com/aws/aws-cdk/pull/14952>.
+
+### Is this a breaking change?
+
+This is not a breaking change.
+
+### What are the drawbacks of this solution?
+
+1. Ideally, the dependencies, `assert`, `cloudformation-diff` and `cfnspec`
+   are all available to be consumed as regular dependencies but are not,
+   since they are part of the same monorepo.
+
+   Instead, the design 'vendors in' these dependencies, i.e., copies over
+   the source code and rewrites imports, thereby increasing tech debt.
+
+2. Any new feature or bug fix to the new assert module that require changes
+   to the API needs to be performed both in the old assert module and new.
+   This can lead to weird/complicated implementation, since they both follow
+   different design patterns.
+
+3. The 'jest' friendly interfaces that were previously available are no longer
+   available.
+   Users previously using -
+
+   ```ts
+   expect(stack).toHaveResource('Foo::Bar', { ... });
+   ```
+
+   will now have to use
+
+   ```ts
+   const assertions = TemplateAssertions.fromStack(stack);
+   assertions.assertResource('Foo::Bar', { ... });
+   ```
+
+### What are the mitigations and/or long term plans to address these concerns?
+
+Once this module reaches general availability, we will deprecate the old 'assert'
+module. At that point, all of its source code can now be made as part of the new
+'assert' module.
+
+The dependencies on 'cloudformation-diff' and 'cfnspec' will need to be carefully
+evaluated and a decision made on whether to lift-and-shift the relevant code, or
+to move these out of the monorepo as independent modules.
+
+Friendly interfaces for popular testing frameworks such as jest, nunit, junit, etc.
+can be built as separate modules specific to those languages (not jsii modules)
+on top of the new 'assert' library.
+If necessary, this can be owned and maintained by the CDK team, but probably more
+suited for the community to pick up and build as third party modules.
```

Comment:
Ah, I see you already thought about it.

In my opinion, this is something that should be owned by our team. I would add them to the RFC plan.

Reply:
We can always add this later. Let's first assess demand by just shipping what we have.